### PR TITLE
Linux arm gnueabihf

### PR DESCRIPTION
**Pull request by dmorilha-twilio**
@berkus this PR adds a Dockerfile to be able to build for both supported architectures on linux: x86_64 and arm_gnueabihf.

Also, it looks like our manually dependency was removed from bintray and we currently can't build, which makes this PR and its integration into the CI a high priority for us.

Please let me know your concerns and comments. Let's try to address them asap!

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [X] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Review by berkus — CHANGES_REQUESTED** [review]
Sorry, but: 
1. Dockerfile should not be a part of this repo, it should go to docker ci repo. 
2. Please rename the commits according to the guideline [here](https://chris.beams.io/posts/git-commit/)

File: boost.sh [hunk 1]
```diff
@@ -675,8 +675,26 @@ EOF
 generateLinuxUserConfig()
 {
     cat > "$BOOST_SRC/tools/build/src/user-config.jam" <<EOF
-using gcc : : g++ $LINUX_ARCH_FLAGS $EXTRA_LINUX_FLAGS
-: <architecture>x86 <target-os>linux
+using gcc : 6.3.0~x86_64
```
Comment (berkus) [inline boost.sh]: Why are you downgrading to gcc 6.3 here?

File: boost.sh [hunk 2]
```diff
@@ -948,24 +966,34 @@ buildBoost_Linux()
     mkdir -p $OUTPUT_DIR
     echo > ${OUTPUT_DIR}/linux-build.log
 
-    for BITS in 64 32; do
+    for BITS in 32 64; do
```
Comment (berkus) [inline boost.sh]: Why did you reverse the order here?

File: boost.sh [hunk 3]
```diff
@@ -1423,7 +1451,7 @@ fi
 # -DBOOST_SP_USE_SPINLOCK
 # -stdlib=libc++
 
-CPPSTD="-std=c++1z"
+CPPSTD="-std=c++14"
```
Comment (berkus) [inline boost.sh]: We use c++17

File: boost.sh [hunk 4]
```diff
@@ -948,24 +966,34 @@ buildBoost_Linux()
...
             echo Building $VARIANT $BITS-bit Boost for Linux
 
             if [[ $BITS == 64 ]]; then
                 LIBDIR_SUFFIX=x86_64
+                ARCH=x86
+                TOOLSET=gcc-6.3.0~x86_64
+                ABI=sysv
             else
-                LIBDIR_SUFFIX=x86
+                LIBDIR_SUFFIX=arm_gnueabihf
```
Comment (berkus) [inline boost.sh]: Why 32 bit builds are only for arm?

File: boost.sh (hunk 1 above)
Comment (dmorilha-twilio) [reply boost.sh]: the initial idea was to rely on webrtc's clang, unfortunately bjam doesn't seem to cope well with it and I was unable to use clang's assembler to build `context`. GCC 6.3.0 seems to be the stable version for the debian 9 rtc-cpp uses...

File: boost.sh (hunk 2 above)
Comment (dmorilha-twilio) [reply boost.sh]: no specific reason...

File: boost.sh (hunk 3 above)
Comment (dmorilha-twilio) [reply boost.sh]: What is the minimum libstdc++ version required? In general c++14 is a stable iteration, while 17 has a lot of new features... I am all up to stabler rather than newer. Not a real problem in case the libstdc++ supports both...

File: boost.sh (hunk 4 above)
Comment (dmorilha-twilio) [reply boost.sh]: i686 seems to be widely unsupported on linux these days

**Comment by dmorilha-twilio** [comment]
> Dockerfile should not be a part of this repo, it should go to docker ci repo

Could you please point me to where this would be? Thank you

File: boost.sh (hunk 3 above)
Comment (dmorilha-twilio) [reply boost.sh]: Actually @dipankadas and I had problems with `libstdc++.so.7` as most of the distributions out there still feature `libstdc++.so.6`. Why do you need C++17 for, specially to compile a dependency? Couldn't you just compile third party libraries like boost with C++14 and enable C++17 on the compilation units you really use the desired features?

**Comment by berkus** [comment]
> > Dockerfile should not be a part of this repo, it should go to docker ci repo
> 
> Could you please point me to where this would be? Thank you

Should be here https://github.com/cibuilderbot/docker-circleci-linux-android/blob/master/Dockerfile

File: boost.sh (hunk 4 above)
Comment (berkus) [reply boost.sh]: We still build it and i dont want to break the release flow for no reason.

File: boost.sh (hunk 3 above)
Comment (berkus) [reply boost.sh]: We could try, since we dont use binary libraries from boost it shouldnt affect us.

File: boost.sh (hunk 2 above)
Comment (berkus) [reply boost.sh]: Lets keep 64 bits a priority then. We want them built first so we can start consuming/trying them right away.

File: boost.sh (hunk 1 above)
Comment (berkus) [reply boost.sh]: Could it harm to use our default gcc-7 or 8 instead?

Comment (dmorilha-twilio) [reply boost.sh]: It turns out ubuntu 16.04, the one being used at the moment does not support either gcc-7-cross or gcc-8-cross. In order to move to a new cross compiler for arm, the simplest would be to update from ubuntu 16.04 to ubuntu 18.04. What do you think?

File: boost.sh (hunk 3 above)
Comment (dmorilha-twilio) [reply boost.sh]: I checked and we the docker image has `libstdc++.so.6.0.28`. Our rtc-cpp circle-ci image has `libstdc++.so.6.0.22`. Here is a list of gcc and their respective libstdc++.so minimum required version: https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html . I've also compiled some stuff with gcc-8 and `-std=c++17` on  and it ran OK with `libstdc++.so.6.0.22`. So we should be fine either way.

**Comment by dmorilha-twilio** [comment]
we should be all good now, all comments addressed

File: boost.sh [hunk 5]
```diff
@@ -962,14 +974,36 @@ buildBoost_Linux()
                 --prefix="$OUTPUT_DIR" \
                 --libdir="$OUTPUT_DIR/lib/$VARIANT/$LIBDIR_SUFFIX" \
                 address-model=$BITS variant=$VARIANT \
-                optimization=speed \
```
Comment (berkus) [inline boost.sh]: Why did you remove optimization flags here?

File: boost.sh [hunk 6]
```diff
@@ -27,8 +27,6 @@
 # tarball in the current directory and the script will use that.
 #
 #===============================================================================
-set -x
```
Comment (berkus) [inline boost.sh:30]: Would be nice to have `set -x` back.

**Review by berkus — CHANGES_REQUESTED** [review]
last two issues to fix

**Comment by berkus** [comment]
Well, three actually - Please rename the commits according to the guideline [here](https://chris.beams.io/posts/git-commit/) - still stands.

File: boost.sh (hunk 6 above)
Comment (dmorilha-twilio) [reply boost.sh:30]: yes, sorry, removed by accident

**Comment by dmorilha-twilio** [comment]
I don't know what a "good commit message" would be, but you are free to change when merging the PR.

**Review by berkus — APPROVED** [review]
